Ammonia models are now faster!

vlas-sokolov opened the pull request:
So I've been profiling the ammonia model performance in `pyspeckit`, and, long story short, found out that both models and spectral fits can be improved to yield a large speed boost.

Here are the two areas I've addressed so far:

First of all, whenever `ammonia` function gets called, it tries to convert `xarr` into `GHz`. Somehow, this isn't optimized for speed in `astropy`, so in the end it ends up eating up a significant chunck of CPU power. This is fixed in 01142ec, and the `%timeit` results are given below in rows 2 and 3.

Now on to the largest performance boost. There's a `line_names` argument that gets passed to `ammonia` function, that, in turn, gets inherited by `_ammonia_spectrum`, specifying the inversion levels to calculate over. There are a few issues with this:

- [feature, not a bug] Because `line_names` defaults to all lines in (1,1)-(8,8) range, the default model that is registered under `Registry.multifitters['ammonia']` is slower than it should be for cubes that have, for example, only (1,1) and (2,2) lines. This can, of course, be solved by registering a new fitter with `function=ammonia_model(line_names=['oneone', 'twotwo'])` passed to the `add_fitter` method, but... I suspect that not many users figured it out. In fact, when I look over to the some data analysis pipelines for NH3 (e.g. GAS survey repo), they don't really use it. if you check the table below, the second and third columns show the time it takes to run `get_full_model` on the (1,1) and (2,2) ammonia spectrum from a typical GBT spectral setup. Clearly, the `ammonia_model` with matching 
`line_names` wins the race!

- However, selecting a proper `ammonia_model` had no effect on the fitting performance. That was, probably, a bug - the fitting function didn't pick up the model `kwargs` that were stored under `fitter.modelfunc_kwargs`. Therefore, even for `line_names=['oneone', 'twotwo']`, the fitter ran with all the inversion lines up to (8,8) regardless of the ammonia model registered, taking the defaults from the ammonia function. I set up the fitting kwargs to update from model ones in 508d946.

| `faster-ammonia` branch commit         | `get_full_model` for `line_names`=(1,1)-(2,2) | `get_full_model` for `line_names`=(1,1)-(8,8) | Fitting a dummy cube for `line_names`=(1,1)-(2,2) | Fitting a dummy cube for `line_names`=(1,1)-(8,8) |
|----------------------------------------|-----------------------------------------------|-----------------------------------------------|---------------------------------------------------|---------------------------------------------------|
| 71552fd (current master)      | 4.62 ms                                       | 9.76 ms                                       | 51.4 s                                            | 51.4 s                                            |
| 01142ec (condition call to as_unit)  | 3.33 ms                                       | 8.44 ms                                       | 40.8 s                                            | 40.8 s                                            |
| 508d946 (fitter inherits model kwargs) | -                                             | -                                             |  **26.7 s**   | 40.9 s                                            |


So, main takeaway is that fitting ammonia cubes can be twice as fast now in some cases!

@keflavich  I am still working on more improvements, so please don't merge just yet.

keflavich commented:
Nice work!  I'll review when you're done.  These little speedups are a great help.

Yes, re-converting units every time is a bad thing.  So is computing the line profile for unobserved regions.

vlas-sokolov commented:
All right, I wrote a few more improvements. I'm using [this script](https://gist.github.com/vlas-sokolov/af058f7e46a67c77fc796c614b5c859e) for timings and setting up models/cubes/spectra, in case anyone wants to experiment with it.

I haven't yet added the commit (40ad5ed) from `faster-mpfit' branch as it involves making changes into mpfit.py - is it okay to do so or is it an untouchable legacy code? Apart from that, #223 fixes some issues I had with lmfit, and together with changes on this PR there's a nice speedup achievable as well.

Both timings, for modified mpfit and for lmfit, are below:

| Branch name           | latest `master`   | `faster-ammonia` (this PR) | `faster-mpfit` with changes from this PR | `lmfit-properpars` (#223) with changes from this PR |
|-----------------------|----------|--------------------------|----------------------------------------|--------------------------------------------|
| Time of `fiteach` run | 51.6 sec | 27.9 sec                 | 20.9 sec                               | 14.9 sec (with `use_lmfit=True`)             |

Also, I wrote a version of `_ammonia_spectrum` function that uses Einstein summation instead of a for loop over the hyperfines (12963a9). It improves the running time of `get_full_model` and `fiteach(use_lmfit=True)` by ~5%, but for `fiteach(use_lmfit=False)` it degrades the performance by as much as a third. I'll try to figure out why, but for now it's probably better to leave it out of the PR.

keflavich commented:
`mpfit.py` is touchable; pyspeckit contains a fork of that code (and it could probably do with a *lot* of optimizations).

If you make the einstein summation change, include some extra documentation with it (i.e., comments).

vlas-sokolov commented:
Okay, I updated the PR with the mpfit tinkering - two conversions from `sum()` to `np.sum()`. It's possible that `mpfit.py`can be made even faster - I just fixed two lines that the profiler thought to be performance bottlenecks. Meanwhile, I'll try to see if I can find the cause for the strange `einsum` behavior...

@keflavich do you think it's worth investigating automatic selection of `line_names` in ammonia models?
 As I have shown above, proper `line_names` result in a major speed boost, but the problem is that, by default, a registered model doesn't know anything about the `xarr` instance it will act upon. However, checking for valid lines in `xarr` at every point a spectrum is calculated would - I'm guessing here - result in a much slower performance (astropy functions tend to be a bit sluggish).

keflavich commented:
@vlas-sokolov should I merge this?  It looks fine; I had nothing to add in review.

Do you know how to run `line_profiler`?  That's the other way we could try to find places to squeeze out a little more performance.

vlas-sokolov commented:
@keflavich I used `statprof` and `pycallgraph` for profiling, they both seem to do the same thing `line_profiler` does.

I think this can be merged. I haven't had the time to look into the `np.einsum` improvement, but I will leave that branch on GitHub for any potential future improvements.
